### PR TITLE
[sglang-miles] Streamed LoRA weight updates: register RPC, session scope, LoRA stash

### PR DESCRIPTION
**Pull request by yueming-yuan**
Engine side of the miles LoRA weight-sync unification (miles radixark/miles#1360 Step 6): adapter weights stop being a separate RPC family and ride the ordinary weight-update stream; a small control plane handles adapter identity. The legacy `load_lora_adapter_from_tensors` / `load_lora_adapter_from_distributed` wire routes are removed here — miles switches to the stream path in the same merge window.

## What

1. **`/register_lora_adapter`** (`RegisterLoRAAdapterReqInput`: `lora_name`, `config_dict`, `pinned=False`): create-or-refresh an adapter's identity and config under `lora_update_lock`. Implementation = the existing manager entry with empty tensors + upsert (`LoRAManager.register_lora_adapter`). Weights are untouched by the caller: a new adapter starts zeroed, re-registration re-zeroes (a slot's new tenant must not serve its predecessor's weights). Registered adapters are **unpinned** — the pool refills a registered adapter lazily from its CPU copy, and pinning every slot would trip the anti-starvation check. Two registration-time rejections: a streamed adapter has no artifact to reload from, so at `--max-loaded-loras` a **new** name is rejected rather than evicting an LRU one (eviction would lose the only engine-side copy; same-name upserts stay allowed at the cap, and the path-loaded route keeps LRU eviction); and a name containing `:` is rejected, since that character separates adapter from tensor key on the wire.

2. **Session scope: `begin_weight_update(selector, sync_base=True)`**. `sync_base=False` declares an adapter-only session: the base quant unpack (`restore_weight`) is skipped at begin, the re-finalize/`post_load` is skipped at end, and a base tensor arriving in-session is an error. This lets adapter-only syncs on a quantized base avoid a full-model requant round-trip every round.

3. **LoRA split + stash in the stream**: `update_weights_from_{tensor,distributed}` partition incoming named tensors on the `{lora_name}:` prefix (`:` is rejected in adapter names and never occurs in an HF param name; a lora_A/lora_B tensor *without* a prefix is rejected as a sender bug). Base tensors proceed exactly as before; prefixed tensors accumulate in a session-scoped stash. For `load_format="flattened_bucket"` the split runs on the reconstructed tensors; a pure-base bucket keeps the original payload and format (zero behavior change for existing users). A per-rank `LocalSerializedTensor` payload is reduced to this rank's tensor at the stash, since a prefixed tensor never reaches the base loader's `_unwrap_tensor`.

4. **Apply at `end_weight_update`**: each stashed adapter goes through `LoRAManager.load_lora_adapter_from_tensors(ref, tensors, config_dict=None → config from registration, upsert=True)` — reusing validation, name normalization (qkv fusion etc.), CPU-copy update, in-place pool-slot rewrite, and rollback wholesale. Guards, all reported to the caller as a failed `end_weight_update` rather than raised: data for an unregistered adapter fails; optional `expected_lora_checksums` (per-adapter sha256 manifest) is checked for adapter-set equality, per-adapter name-set equality, and values, supporting miles' `--check-lora-weight-equal`; without checksums, an adapter whose streamed tensor-name set differs from its previous sync is rejected (a partial stream would otherwise be silently zero-filled by the whole-adapter replace), and re-registering or unloading a name clears that guard, since it is a new adapter identity. The stash is session-scoped: `begin` discards leftovers from an aborted round, `end` applies and clears.

## Not in this PR

- miles-side switch (iterator emits prefixed adapter units; deletes its `send_adapter` path) — lands together with the version bump.

## Tests

- `test/registered/unit/lora/test_lora_stream_update.py` (CPU, mock-based, same style as `test_lora_upsert.py`): split guard, stash grouping, per-rank unwrap, checksum/name-set/adapter-set verification, the partial-stream guard and its reset on re-register/unload, error propagation, stash lifecycle, and the manager entry points.
- `test/registered/rl/test_lora_load_from_tensor.py` (GPU): stream load/unload/reload, cap rejection, separator rejection, flattened-bucket path, and log-prob parity against HuggingFace PEFT.

---
### CI States

Latest PR Test (Base): :x: [Run #33593225648](https://github.com/sgl-project/sglang/actions/runs/33593225648)
Latest PR Test (Extra): :x: [Run #33593225420](https://github.com/sgl-project/sglang/actions/runs/33593225420)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33593225582](https://github.com/sgl-project/sglang/actions/runs/33593225582)

### CODE REVIEW

**Comment by yueming-yuan** [comment]
Second commit folds the legacy-route removal into this PR (per the simultaneous-merge plan with radixark/miles#2756): deletes the load_lora_adapter_from_tensors / load_lora_adapter_from_distributed wire routes across all layers (−539 lines; the LoRAManager entry point stays as the stream-apply implementation), migrates test_lora_load_from_tensor.py onto register + sync_base=False stream, repurposes the multi-tokenizer-worker upsert guard onto register_lora_adapter, and fixes two gaps in the first commit (missing tokenizer-level register handler and scheduler-level method). Manager-level tests (test_lora_upsert, test_lora_stream_update) are unaffected. **Merge this and #2756 together; there is no compat window.**

**Review by yushengsu-thu — APPROVED** [review]

**Review by jybsuper — COMMENTED** [review]
Left a few inline comments (lint, two broken tests, a default mismatch, and one state-cleanup issue in the new stash guard).

Suggest adding a few targeted tests once CI can run:
- Fix the tuple bug and get the existing e2e file green — it already covers register → stream → apply → generate (incl. flattened bucket).
- Unit test: unload/re-register a name, stream a *different* tensor set, assert it's accepted (pins the `_lora_applied_names` fix).

File: python/sglang/srt/managers/tp_worker.py [hunk 1]
```diff
@@ -53,14 +52,12 @@
 from sglang.srt.model_executor.pool_configurator import MemoryPoolConfig
 from sglang.srt.runtime_context import get_exec, get_model, get_schedule, get_spec
 from sglang.srt.server_args import ServerArgs
-from sglang.srt.utils import MultiprocessingSerializer, broadcast_pyobj, set_random_seed
+from sglang.srt.utils import broadcast_pyobj, set_random_seed
```
Comment (jybsuper) [inline python/sglang/srt/managers/tp_worker.py:55]: lint is red: `_deserialize_own_rank` below lost its imports (`MultiprocessingSerializer`, `monkey_patch_torch_reductions`) and its only caller in this PR — please delete the dead helper (pre-commit/ruff F821).

File: test/registered/rl/test_lora_load_from_tensor.py [hunk 2]
```diff
@@ -22,6 +22,21 @@
...
+    result = engine.register_lora_adapter(lora_name=lora_name, config_dict=config_dict)
+    if not result.success:
+        return result
+    engine.begin_weight_update(selector="all", sync_base=False)
+    engine.update_weights_from_tensor(
+        named_tensors=[
+            (f"{lora_name}:{key}", tensor) for key, tensor in tensors.items()
+        ]
+    )
+    return engine.end_weight_update()
```
Comment (jybsuper) [inline test/registered/rl/test_lora_load_from_tensor.py]: `Engine.end_weight_update()` returns the `(success, message)` tuple from `merge_results`, but `load_lora_via_stream` callers do `result.success` / `result.error_message` — every test in this file dies with `AttributeError` on its first assertion. Please unpack the tuple (or wrap the return), and note `result.loaded_adapters` no longer exists on any end_weight_update reply.

File: python/sglang/srt/managers/tokenizer_control_mixin.py [hunk 3]
```diff
@@ -720,145 +718,48 @@ async def load_lora_adapter(
...
-        if obj.upsert and self.server_args.tokenizer_worker_num > 1:
+        if self.server_args.tokenizer_worker_num > 1:
             raise ValueError(
                 "LoRA upsert is not supported with tokenizer_worker_num > 1: "
                 "each HTTP worker resolves lora_name against its own registry, "
                 "making upsert nondeterministic across workers."
             )
 
-    async def load_lora_adapter_from_tensors(
+    async def register_lora_adapter(
```
Comment (jybsuper) [inline python/sglang/srt/managers/tokenizer_control_mixin.py:745]: The new register path drops the `max_loaded_loras` LRU enforcement that both legacy routes had — intentional? If streamed adapters are meant to be exempt (trainer controls the count), I'd suggest rejecting registration over the cap instead of ignoring it (evicting a wire-loaded adapter is destructive), and `test_lora_lru_eviction` needs a rewrite either way since it asserts eviction that no longer exists.

File: python/sglang/srt/entrypoints/engine.py [hunk 4]
```diff
@@ -1535,67 +1550,6 @@ def _serialize_tensors_per_rank(
...
-            lora_name=lora_name,
-            lora_path=lora_path,
-            pinned=pinned,
-        )
-
-        return self.loop.run_until_complete(
-            self.tokenizer_manager.load_lora_adapter(obj, None)
-        )
-
     def unload_lora_adapter(self, lora_name: str):
```
Comment (jybsuper) [inline python/sglang/srt/entrypoints/engine.py:1566]: Deleting sync `Engine.load_lora_adapter` breaks `test/registered/lora/test_lora_eviction.py`: `SRTRunner.load_lora_adapter` (runners.py:663) now resolves to the 2-arg `EngineBase` `pass` stub and raises `TypeError` on the 3-arg call. Please keep a sync wrapper (e.g. over `async_load_lora_adapter`) or update runners/EngineBase together.

File: python/sglang/srt/managers/scheduler_components/weight_updater.py [hunk 5]
```diff
@@ -342,12 +408,61 @@ def end_weight_update(self, recv_req: EndWeightUpdateReqInput):
...
+                    return False, (
+                        f"[LORA-CHECK] adapter {lora_name!r}: streamed tensor names "
+                        f"do not match the expected manifest"
+                    )
+                for name in sorted(tensors):
+                    if _sha256_tensor(tensors[name]) != expected[name]:
+                        return False, (
+                            f"[LORA-CHECK] adapter {lora_name!r}: checksum mismatch for {name!r}"
+                        )
+            prev = self._lora_applied_names.get(lora_name)
```
Comment (jybsuper) [inline python/sglang/srt/managers/scheduler_components/weight_updater.py:482]: The new `_lora_applied_names` guard is a good addition, but nothing ever clears it: re-registering a name with a different config — e.g. different `target_modules`, exactly the "new tenant" flow register documents — gets permanently rejected as "different tensor set … (partial stream?)" until restart. Can the register/unload handlers clear the name's entry (or key this by `lora_id` instead of name)?

File: python/sglang/srt/managers/tokenizer_control_mixin.py (hunk 3 above)
Comment (yueming-yuan) [reply python/sglang/srt/managers/tokenizer_control_mixin.py:745]: good catch, now reject the overflow adaptor in stream path (keep reload them in normal load path), test added.

**Comment by yueming-yuan** [comment]
/rerun-test test_lora_stream_update.py test_lora_upsert.py test_lora_lease.py test_msgpack_ipc_roundtrip.py test_lora_load_from_tensor.py test_lora_eviction.py

File: python/sglang/srt/managers/scheduler_components/weight_updater.py [hunk 6]
```diff
@@ -342,12 +411,70 @@ def end_weight_update(self, recv_req: EndWeightUpdateReqInput):
...
+            ), f"streamed LoRA tensor {prefixed_name!r} must arrive as a plain tensor"
+            self._lora_stash.setdefault(lora_name, {})[hf_key] = tensor
+
+    def _apply_lora_stash(
+        self, expected_checksums: Optional[Dict[str, Dict[str, str]]]
+    ) -> Tuple[bool, str]:
+        """Hand each streamed adapter to the LoRA manager's whole-adapter entry
+        point (config from registration, upsert in place), then clear the stash."""
+        if expected_checksums is not None:
+            assert set(expected_checksums) == set(self._lora_stash)
```
Comment (jybsuper) [inline python/sglang/srt/managers/scheduler_components/weight_updater.py]: This assert replaces the previous graceful `expected is None` path: a manifest/stash adapter-set mismatch now kills the scheduler (`end_weight_update` has no try/except and the dispatcher has no catch-all) instead of returning a `[LORA-CHECK]` failure like the per-adapter checks right below it. Reachable from a sender bug — e.g. miles pre-seeds the manifest with every adapter it plans to push, so an adapter that ends up streaming zero tensors would trip it. Could this return `(False, msg)` naming the two sets instead?

**Review by jybsuper — COMMENTED** [review]
Thanks for the quick turnaround — the fixes look right, and I verified them on a devbox since CI can't run on this base: the four CPU unit files pass (the 3 remaining failures fail identically on the base commit), `test_lora_load_from_tensor.py` passes 6/6 on GPU, and the restored sync `Engine.load_lora_adapter` loads/serves/unloads an adapter with the 3-arg call `SRTRunner` uses.

One small ask on the description: it still says "Pure addition — the existing `load_lora_adapter_from_tensors` / `load_lora_adapter_from_distributed` wire routes are untouched", `pinned=True`, and "pinned by default" — since the body becomes the squash-merge message, could you refresh it to match the current diff (routes removed per the simultaneous-merge plan, unpinned default, cap rejection)?

File: test/registered/unit/lora/test_lora_stream_update.py [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+        result = manager.apply_streamed_adapter("missing", {LORA_A: torch.zeros(1)})
+        self.assertFalse(result.success)
+        self.assertIn("unregistered", result.error_message)
+
+
+if __name__ == "__main__":
+    unittest.main()
+
+
+class TestWeightVersionDeferredToCommit(CustomTestCase):
```
Comment (jybsuper) [inline test/registered/unit/lora/test_lora_stream_update.py:270]: nit: this class is defined after `unittest.main()`, and CI runs files as `python3 file.py -f`, so these 3 tests never execute in CI (main() exits before the class exists). Move it above the `__main__` block.

**Review by jybsuper — COMMENTED** [review]
All other comments are addressed and everything else looks good to me. Re-verified the latest head on a devbox: the CPU unit files pass (64 tests, the 3 remaining failures are pre-existing on the base commit) and `test_lora_load_from_tensor.py` passes 6/6 on GPU. Only one nit left inline on the test file placement.
